Mathlib/FieldTheory/SeparablyGenerated.lean

```diff
@@ -0,0 +1,321 @@
+  by_contra!
+  have (σ) (hσ : σ ∈ F.support) : ∃ σ', σ = p • σ' := by
+    choose σ' hσ' using (this · σ hσ)
+    exact ⟨⟨σ.support, σ', by simp [hσ', hp.ne_zero]⟩, Finsupp.ext hσ'⟩
+  choose! σ' hσ' using this
+  have hσ'' (σ : F.support) : σ.1 = p • σ' σ := hσ' σ.1 σ.2
+  classical
+  replace H (ι : Type u_3) (_ : Fintype ι) (v : ι → K) (hv : LinearIndependent k v) :
+      LinearIndependent k (v · ^ p) := by
+    simpa only [Finset.coe_image, Finset.coe_univ, Set.image_univ, linearIndepOn_range_iff
+      hv.injective] using H (Finset.univ.image v) (by simpa using hv.linearIndepOn_id)
+  have := mt (H F.support inferInstance (fun s ↦ aeval a (monomial (σ' s) (1 : k)))) (by
+    simp_rw [← map_pow, monomial_pow, ← hσ'', one_pow, not_linearIndependent_iff]
+    refine ⟨.univ, (F.coeff ·), ?_, by simpa [MvPolynomial.eq_zero_iff] using hF0⟩
+    simp only [← map_smul, ← map_sum, Finset.univ_eq_attach, smul_eq_mul, mul_one]
+    rw [F.support.sum_attach (fun i ↦ monomial i (F.coeff i)), support_sum_monomial_coeff, hFa])
+  simp only [LinearIndependent, injective_iff_map_eq_zero, not_forall] at this
+  obtain ⟨F', hF', hF'0⟩ := this
+  let F'' : MvPolynomial ι k := F'.mapDomain fun s ↦ σ' s.1
+  have hF''0 : F'' ≠ 0 := ne_of_ne_of_eq ((Finsupp.mapDomain_injective fun s t h ↦ Subtype.ext
+    (Finsupp.ext fun i ↦ by rw [hσ' _ s.2, hσ' _ t.2, h])).ne_iff.mpr hF'0) (by simp)
+  have hF'' : aeval a F'' = 0 := by
+    have : (aeval a).toLinearMap ∘ₗ (Finsupp.lmapDomain k k fun s : F.support ↦ σ' s) =
+        (Finsupp.linearCombination k fun s : F.support ↦ aeval a (monomial (σ' s) (1 : k))) := by
+      ext v; simp [MvPolynomial, AddMonoidAlgebra, monomial]
+    simp only [← hF', F'', ← this]; rfl
+  suffices hpm : p * F''.totalDegree ≤ F.totalDegree by
+    have hF''0' : F''.totalDegree ≠ 0 := by
+      contrapose! hF''0
+      rw [totalDegree_eq_zero_iff_eq_C.mp hF''0, aeval_C, map_eq_zero] at hF''
+      rw [totalDegree_eq_zero_iff_eq_C.mp hF''0, hF'', map_zero]
+    replace this := hpm.trans ((HF F'' hF''0 hF'').trans_eq (one_mul _).symm)
+    exact hp.one_lt.not_ge ((mul_le_mul_iff_of_pos_right hF''0'.bot_lt).mp this)
+  rw [totalDegree, Finset.mul_sup₀, Finset.sup_le_iff]
+  intro σ hσ
+  obtain ⟨σ, hσ₂, rfl⟩ := Finset.mem_image.mp (Finsupp.mapDomain_support hσ)
+  refine le_trans ?_ (Finset.le_sup σ.2)
+  conv_rhs => rw [hσ' _ σ.2, Finsupp.sum_smul_index (fun _ ↦ rfl), ← Finsupp.mul_sum]
+
+end MvPolynomial
+
+open IntermediateField
+
+section
+
+variable [ExpChar k p]
+
+include hp H
+
+/--
+Suppose `k` has chararcteristic `p` and `a₁,...,aₙ` is a transcendental basis of `K/k`.
+Suppose furthermore that if `{ sᵢ } ⊆ K` is an arbitrary `k`-linearly independent set,
+`{ sᵢᵖ } ⊆ K` is also `k`-linearly independent (which is true when `K ⊗ₖ k^{1/p}` is reduced).
+
+Then some subset of `a₁,...,aₙ₊₁` forms a transcedental basis that `a₁,...,aₙ₊₁` are separable over.
+-/
+lemma exists_isTranscendenceBasis_and_isSeparable_of_linearIndepOn_pow
```

Comment:
Here my profiler tells me that aesop takes a second (and the infoview tells me that you're proving the same goal thrice in `ha''`. Is this worth extracting, at least as a have?